app/furniture/marketplace/marketplace_component.rb

```diff
@@ -14,7 +14,7 @@ def render?
 
     def cart
       delivery_area = marketplace.delivery_areas.first if marketplace.delivery_areas.size == 1
-      @cart ||= carts.create_with(contact_email: shopper.person&.email).find_or_create_by(delivery_area: delivery_area, shopper: shopper, status: :pre_checkout)
+      @cart ||= carts.create_with(contact_email: shopper.person&.email, delivery_area: delivery_area).find_or_create_by(shopper: shopper, status: :pre_checkout)
```

Comment:
I think we should only have 1 cart per (marketplace, shopper) pair, no matter what is the chosen delivery area, so the delivery area shouldn't be mentioned here at all.

From a user perspective, I don't expect an online store to keep for me two separate carts depending on where my order is getting delivered to. I expect to have 1 cart, and then I can change the delivery area (and delivery address) before checking out as needed.

Reply:
Agreed; that's why there was the bug, it was trying to create a cart each time; rather than including the default `delivery_area` in cases where there is only 1 for the marketplace.

Now it only has 1 cart per shopper per marketplace; since the `find_or_create_by` no longer references the `delivery_area`.

Reply:
Ah, yes, d'oh, I was reading that the other way around.

Reply:
I can imagine splitting the `create_with` and `find_or_create_by` onto different lines would help 🙈

Reply:
I think I hate `find_or_create_by`. This is a huge conflation of operations. 😅

Reply:
It is extra-fun if you chain the find-or-create with a create-with! 😹

Reply:
If we wanted to stop using both of those, we would need to write something like this:

```rb
cart = carts.find_by(shopper:, status: :pre_checkout) || carts.create(shopper: shopper, status: :pre_checkout)
cart.update(contact_email: shopper.person&.email,  delivery_area:) if cart.previously_new_record?
```

Which I don't find much more legible than:

```rb
cart = carts.create_with(contact_email: shopper.person&.email,  delivery_area:)
            .find_or_create_by(shopper:, status: :pre_checkout)
```

I agree mashing them both on one line is a bit of a disaster; but both `find_or_create_by` and `create_with` are useful tools.

Reply:
🏌️ Is it un-chill to do something like this in ruby/Rails? Is there supposed to be a perf gain or something with those convenience query methods? (Is this statement below materially different?)

```ruby
cart = carts.find_by(shopper:, status: :pre_checkout)
if cart.blank?
  cart = carts.create(
    shopper: shopper, 
    status: :pre_checkout, 
    contact_email: shopper.person&.email, 
    delivery_area:
)
```

Reply:
I think that also works tho the re-assignment in an if block feels odd to me but not in a horrible way.

Reply:
I personally find Ross's snippet easier to read and understand. Here is a riff on that, but without the re-assignment:
```ruby
cart = carts.find_by(shopper:, status: :pre_checkout) || carts.create(
  shopper: shopper, 
  status: :pre_checkout, 
  contact_email: shopper.person&.email, 
  delivery_area:
)
```

Reply:
And to wrap it all up, this could be moved into a factory/finder method on `Marketplace`, something like `Marketplace.find_or_create_cart(shopper:, delivery_area: nil)`.

Reply:
🚢 :it:

Reply:
https://github.com/zinc-collective/convene/pull/2009/